src/arrayMethodJoin.js

```diff
@@ -5,7 +5,51 @@
  */
 function applyCustomJoin() {
   [].__proto__.join2 = function(separator) {
-    // write code here
+    if (this.length === 0) {
+      return '';
+    }
+
+    if (this.length === 1) {
+      return this[0] + '';
+    }
+
+    let string = '';
+
+    if ((separator === undefined)) {
+      for (let i = 0; i < this.length; i++) {
+        if (i !== this.length - 1) {
+          string += this[i] + ',';
+        } else {
+          string += this[i];
+        }
+      }
+    }
+
+    if (typeof separator === 'string') {
+      for (let i = 0; i < this.length; i++) {
+        if ((this[i] === undefined) || (this[i] === null)) {
+          this[i] = '';
+        }
+
+        if (i !== this.length - 1) {
+          string += this[i] + `${separator}`;
+        } else {
+          string += this[i];
+        }
+      }
+    }
+
+    if (typeof separator === 'object') {
+      for (let i = 0; i < this.length; i++) {
+        if (i !== this.length - 1) {
+          string += this[i] + `${separator}`;
+        } else {
+          string += this[i];
+        }
```

Comment:
Don't repeat yourself. You have too much similar code blocks. Think about how you can simplify it.